### PR TITLE
KAFKA-5817: Add Serialized class and overloads to KStream#groupBy and KStream#groupByKey

### PR DESCRIPTION
**Pull request by dguy**
Part of KIP-182
- Add the `Serialized` class
- implement overloads of `KStream#groupByKey` and KStream#groupBy`
- deprecate existing methods that have more than default arguments

### CODE REVIEW

**Comment by dguy** [comment]
@guozhangwang, @mjsax, @bbejeck

File: docs/streams/developer-guide.html [hunk 1]
```diff
@@ -833,15 +834,17 @@
...
                        );
 
                        // Group the table by a new key and key type, and also modify the value and value type.
                        KGroupedTable&lt;String, Integer&gt; groupedTable = table.groupBy(
                            (key, value) -> KeyValue.pair(value, value.length()),
-                           Serdes.String(), /* key (note: type was modified) */
-                           Serdes.Integer() /* value (note: type was modified) */
+                           Serialized.with(
+                            Serdes.String(), /* key (note: type was modified) */
+                            Serdes.Integer()) /* value (note: type was modified) */
```
Comment (bbejeck) [inline docs/streams/developer-guide.html]: nit: indent the same as others above

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 2]
```diff
@@ -1012,14 +1012,40 @@ void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
...
+     * {@link #groupByKey(Serialized)} instead.
      *
      * @return a {@link KGroupedStream} that contains the grouped records of the original {@code KStream}
      * @see #groupBy(KeyValueMapper)
      */
     KGroupedStream<K, V> groupByKey();
 
     /**
+     * Group the records by their current key into a {@link KGroupedStream} while preserving the original values
+     * and using the serializers as defined by {@link Serialized}
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: nit: missing . at end

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 3]
```diff
@@ -1012,14 +1012,40 @@ void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
...
+     * {@link #through(String)}) an internal repartitioning topic will be created in Kafka.
+     * This topic will be named "${applicationId}-XXX-repartition", where "applicationId" is user-specified in
+     * {@link StreamsConfig} via parameter {@link StreamsConfig#APPLICATION_ID_CONFIG APPLICATION_ID_CONFIG}, "XXX" is
+     * an internally generated name, and "-repartition" is a fixed suffix.
+     * You can retrieve all generated internal topic names via {@link KafkaStreams#toString()}.
+     * <p>
+     * For this case, all data of this stream will be redistributed through the repartitioning topic by writing all
+     * records to it, and rereading all records from it, such that the resulting {@link KGroupedStream} is partitioned
+     * correctly on its key.
+     * If the last key changing operator changed the key type, it is recommended to use
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: missing end of sentence  or  maybe remove this line altogether?

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 4]
```diff
@@ -1074,6 +1102,32 @@ void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
     <KR> KGroupedStream<KR, V> groupBy(final KeyValueMapper<? super K, ? super V, KR> selector);
 
     /**
+     * Group the records of this {@code KStream} on a new key that is selected using the provided {@link KeyValueMapper}
+     * and {@link Serde}s as specified by {@link Serialized}
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: nit: missing . at end

File: streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java [hunk 5]
```diff
@@ -1074,6 +1102,32 @@ void process(final ProcessorSupplier<? super K, ? super V> processorSupplier,
...
+     * Grouping a stream on the record key is required before an aggregation operator can be applied to the data
+     * (cf. {@link KGroupedStream}).
+     * The {@link KeyValueMapper} selects a new key (with should be of the same type) while preserving the original values.
+     * If the new record key is {@code null} the record will not be included in the resulting {@link KGroupedStream}
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java]: nit: missing . at end

File: streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+ *
+ * @param <K> the key type
+ * @param <V> the value type
+ */
+public class Serialized<K, V> {
+
+    private Serde<K> keySerde;
+    private Serde<V> valueSerde;
+
+    private Serialized(final Serde<K> keySerde, final Serde<V> valueSerde) {
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java]: nit: single parameter per line

File: streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+        return keySerde;
+    }
+
+    public Serde<V> valueSerde() {
+        return valueSerde;
+    }
+
+    /**
+     * Construct a {@code Serialized} instance with the provided key and value {@link Serde}s
+     * If the {@link Serde} params are null the default serdes defined in the configs will be used
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java]: nit: both lines missing . at end

File: streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+     *
+     * @param keySerde   keySerde that will be used to materialize a stream
+     *                   if not specified the default serdes defined in the configs will be used
+     * @param valueSerde valueSerde that will be used to materialize a stream
+     *                   if not specified the default serdes defined in the configs will be used
+     * @param <K>        the key type
+     * @param <V>        the value type
+     * @return a new instance of {@code Serialized} configured with the provided serdes
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java]: why `@code` vs `@link` ?

File: streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+     */
+    public static <K, V> Serialized<K, V> with(final Serde<K> keySerde, final Serde<V> valueSerde) {
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java]: nit: single parameter per line

File: streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+        return new Serialized<>(keySerde, valueSerde);
+    }
+
+    /**
+     * Construct a {@code Serialized} instance with the provided key {@link Serde}
+     * If the {@link Serde} params are null the default serdes defined in the configs will be used
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java]: same as above, missing . at end

File: streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+     *                 if not specified the default serdes defined in the configs will be used
+     * @return a new instance of {@code Serialized} configured with the provided key serde
+     */
+    public Serialized<K, V> withKeySerde(final Serde<K> keySerde) {
+        return new Serialized<>(keySerde, null);
+    }
+
+    /**
+     * Construct a {@code Serialized} instance with the provided value {@link Serde}
+     * If the {@link Serde} params are null the default serdes defined in the configs will be used
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java]: ditto

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java [hunk 12]
```diff
@@ -153,7 +154,7 @@ public void testJoin() throws Exception {
 
         KStream<String, String> stream1 = builder.stream(strSerde, strSerde, topic1);
         KTable<Windowed<String>, String> table1 =
-            stream1.groupByKey(strSerde, strSerde)
+            stream1.groupByKey(Serialized.with(strSerde, strSerde))
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java:156]: Can we format this `Serialized.with` like the one above?

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java [hunk 13]
```diff
@@ -164,7 +165,7 @@ public void testJoin() throws Exception {
 
         KStream<String, String> stream2 = builder.stream(strSerde, strSerde, topic2);
         KTable<Windowed<String>, String> table2 =
-            stream2.groupByKey(strSerde, strSerde)
+            stream2.groupByKey(Serialized.with(strSerde, strSerde))
```
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java:167]: ditto here

File: docs/streams/developer-guide.html [hunk 14]
```diff
@@ -833,15 +834,17 @@
...
+                                Serdes.String())  /* value */
                        );
 
                        // Group the table by a new key and key type, and also modify the value and value type.
                        KGroupedTable&lt;String, Integer&gt; groupedTable = table.groupBy(
                            (key, value) -> KeyValue.pair(value, value.length()),
-                           Serdes.String(), /* key (note: type was modified) */
-                           Serdes.Integer() /* value (note: type was modified) */
+                           Serialized.with(
+                            Serdes.String(), /* key (note: type was modified) */
```
Comment (mjsax) [inline docs/streams/developer-guide.html]: Nit: fix indention

File: streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java (hunk 6 above)
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java]: Nit: formatting

File: streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java (hunk 9 above)
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java]: nit: formatting

File: streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java (hunk 7 above)
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java]: nit: `null` -> {@code null}`

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java]: Nit `.` at the end

File: streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java (hunk 10 above)
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java]: `.`

File: streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java [hunk 15]
```diff
@@ -0,0 +1,86 @@
...
+     * @param <K>        the key type
+     * @param <V>        the value type
+     * @return a new instance of {@code Serialized} configured with the provided serdes
+     */
+    public static <K, V> Serialized<K, V> with(final Serde<K> keySerde, final Serde<V> valueSerde) {
+        return new Serialized<>(keySerde, valueSerde);
+    }
+
+    /**
+     * Construct a {@code Serialized} instance with the provided key {@link Serde}
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java]: `.`

File: streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+     * @param keySerde keySerde that will be used to materialize a stream
+     *                 if not specified the default serdes defined in the configs will be used
+     * @return a new instance of {@code Serialized} configured with the provided key serde
+     */
+    public Serialized<K, V> withKeySerde(final Serde<K> keySerde) {
+        return new Serialized<>(keySerde, null);
+    }
+
+    /**
+     * Construct a {@code Serialized} instance with the provided value {@link Serde}
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java]: `.`

File: streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java (hunk 11 above)
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java]: `.`

File: streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+     * If the {@link Serde} params are null the default serdes defined in the configs will be used
+     *
+     * @param valueSerde keySerde that will be used to materialize a stream
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java]: `keySerde` -> `valueSerde`

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java [hunk 18]
```diff
@@ -714,36 +715,50 @@ public boolean test(final K1 key, final V1 value) {
...
+                                              final Serialized<KR, V> serialized) {
         Objects.requireNonNull(selector, "selector can't be null");
+        Objects.requireNonNull(serialized, "serialized can't be null");
         String selectName = internalSelectKey(selector);
         return new KGroupedStreamImpl<>(builder,
                                         selectName,
                                         sourceNodes,
-                                        keySerde,
-                                        valSerde, true);
+                                        serialized.keySerde(),
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:730]: I am wondering, if we should to even more refactoring an pass in `Serialized` directly here to reduce the number of parameters. Not sure though if the win out weights the refactoring effort. Same for the other PRs btw. Any thoughts? \cc @guozhangwang @bbejeck

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java [hunk 19]
```diff
@@ -714,36 +715,50 @@ public boolean test(final K1 key, final V1 value) {
...
+                                        serialized.valueSerde(), true);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java]: nit: `true` should go to new line.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java [hunk 20]
```diff
@@ -60,8 +61,8 @@ public void testAggBasic() throws Exception {
 
         KStream<String, String> stream1 = builder.stream(strSerde, strSerde, topic1);
         KTable<Windowed<String>, String> table2 =
-            stream1.groupByKey(strSerde,
-                               strSerde)
+            stream1.groupByKey(Serialized.with(strSerde,
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java]: nit: to get the formatting with a class, put into a single line as below.

**Review by mjsax — CHANGES_REQUESTED** [review]
Most is nitpick...

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java (hunk 12 above)
Comment (dguy) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java:156]: i'll just make them like this one as it fits comfortably on a single line

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java (hunk 18 above)
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:730]: I did consider the same. Though was on the fence with it so left it as is

File: streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+    private Serde<K> keySerde;
+    private Serde<V> valueSerde;
+
+    private Serialized(final Serde<K> keySerde,
+                       final Serde<V> valueSerde) {
+        this.keySerde = keySerde;
+        this.valueSerde = valueSerde;
+    }
+
+    public Serde<K> keySerde() {
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java:39]: See my other comment about hiding these getters from users.

**Review by guozhangwang — COMMENTED** [review]
LGTM

File: streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java (hunk 21 above)
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java:39]: I'm going to merge this as is. If we can think of a way to remove these that will work, then i'll file another JIRA to do this.

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/Serialized.java:39]: Sounds good. Thanks @dguy

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java (hunk 18 above)
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java:730]: My preference is to reduce the number of parameters where we can, especially for consistency WRT KIP-182, but I don't have a strong opinion, so I'd be okay if we left it as well.

**Pull request closed without merging**